ecs-agent/acs/session/task_manifest_responder_test.go

```diff
@@ -0,0 +1,149 @@
+	mockSNA.EXPECT().GetLatestSequenceNumber()
+	mockSNA.EXPECT().SetLatestSequenceNumber(testManifest).Return(nil)
+
+	// After we get the latest sequence number, we will expect to set the latest message ID.
+	mockIDA.EXPECT().SetMessageID(testconst.MessageID).Return(nil)
+
+	// Test Metric publication when our responder starts and finishes message processing.
+	mockEntry := mock_metrics.NewMockEntry(ctrl)
+	mockMetrics.EXPECT().New(metrics.TaskManifestHandlingDuration).Return(mockEntry)
+	mockEntry.EXPECT().Done(nil).AnyTimes()
+
+	// When the client queries tasks to be stopped, inject this task to be stopped.
+	mockComparer.EXPECT().CompareRunningTasksOnInstanceWithManifest(testManifest).Return(
+		[]*ecsacs.TaskIdentifier{
+			{DesiredStatus: aws.String("STOPPED"), TaskArn: aws.String(testconst.TaskARN)},
+		},
+		nil,
+	)
+
+	// Handle the task manifest message update.
+	handleManifestMessage := testTaskManifestResponder.HandlerFunc().(func(*ecsacs.TaskManifestMessage))
+	handleManifestMessage(testManifest)
+}
+
+// TestTaskManifestStaleMessage will make sure the manifest can update, retrieve and catch stale
+// Manifest sequence numbers appropriately.
+func TestTaskManifestStaleMessage(t *testing.T) {
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	testResponseSender := func(response interface{}) error {
+		return nil
+	}
+
+	mockComparer := mock_session.NewMockTaskComparer(ctrl)
+	mockIDA := mock_session.NewMockManifestMessageIDAccessor(ctrl)
+	mockSNA := mock_session.NewMockSequenceNumberAccessor(ctrl)
+
+	testTaskManifestResponder := NewTaskManifestResponder(mockComparer, mockSNA, mockIDA, metrics.NewNopEntryFactory(), testResponseSender)
+
+	// Create test task manifest.
+	testManifest := setupTestManifestMessage()
+
+	// Set up a new manifest with a stale number and distinct message ID.
+	newManifest := &ecsacs.TaskManifestMessage{
+		ClusterArn:           aws.String(testconst.ClusterName),
+		ContainerInstanceArn: aws.String(testconst.ContainerInstanceARN),
+		MessageId:            aws.String("456"),
+		Tasks:                []*ecsacs.TaskIdentifier{},
+		Timeline:             aws.Int64(EndingTaskManifestSequenceNumber),
+	}
+
+	// Expect GetLatestSequenceNumber once in handleManifestMessage, allowing it to pass
+	// by comparing it against a lower sequence number.
+	mockSNA.EXPECT().GetLatestSequenceNumber().Return(int64(EndingTaskManifestSequenceNumber)).AnyTimes()
+
+	// The test manifest should be valid, updating sequence number and message ID.
+	mockSNA.EXPECT().SetLatestSequenceNumber(testManifest)
+	mockIDA.EXPECT().SetMessageID(testconst.MessageID)
+	mockComparer.EXPECT().CompareRunningTasksOnInstanceWithManifest(testManifest).Return([]*ecsacs.TaskIdentifier{}, nil)
+
+	handleManifestMessage := testTaskManifestResponder.HandlerFunc().(func(*ecsacs.TaskManifestMessage))
+
+	// Handle the task manifest message update, this should correctly set the message ID and sequence number.
```

Comment:
> this should correctly set the message ID and sequence number.

We should assert this, right?

Reply:
I think this is asserted with our mock behavior when we expect a call to setmessageID and setsequencenumber with the right vals.

Reply:
My understanding is that the mock behavior tells us that we should expect such calls to be made, but doesn't actually assert what those calls did/accomplished.

Reply:
When we assert something like `EXPECT().SetSequenceNumber()`, we will expect the task manifest responder to be taking our values passed into `handleManifestMessage`, checking for staleness, and calling `SetSequenceNumber()`. The second time we call `handleManifestMessage`, the getter and setter are implicitly tested by doing that stale check again. The setter must have worked the first time to correctly set sequence num, our getter to return the right value to compare them, and finally the lack of calls to `SetMessageID` and `SetSequenceNum` with a stale message validate their behavior.

Reply:
> finally the lack of calls to SetMessageID and SetSequenceNum with a stale message validate their behavior

Got it, makes sense. Thanks for emphasizing the called "only once" in the latest revisions!